Merge Decoration class.

tclaar opened the pull request:
Add Decoration class (decoration.js).
Add spritesheets for mayor's house, Chad's house, and tree.

b1gd3vd0g committed:
Merge pull request #70 from GoodBadChad/merge

Merge

tclaar committed:
Added Decoration class and a number of spritesheets.
Added Decoration class (decoration.js).
Added blacksmith's house, Chad's house, mayor's house, and tree spritesheets.

tclaar committed:
Remove troublesome spritesheet.

tclaar committed:
Merge branch 'main' of https://github.com/tclaar/good-bad-chad-trae
 Lines starting with '#' will be ignored, and an empty message aborts

tclaar committed:
Merge pull request #79 from tclaar/main

Push Decoration class.

b1gd3vd0g commented on js/classes/entities/background/decoration.js:
```diff
@@ -0,0 +1,88 @@
+     * @param {Vector} pos the position of the bottom left corner of the Decoration (in world space) 
+     */
+    constructor(type, pos) {
+        const properties = Decoration.PROPERTY_TABLE[type];
```
May be a good idea to throw an error/exception here, with a clear explanation, when an improper "type" is passed in.

calebkrauter replied:
I agree. I assume type would be an int and there would be constants like CHAD_HOUSE or such and such for the keys.

tclaar replied:
Throwing an exception for an invalid argument is a good idea, thanks guys. @calebkrauter I'm not so sure that named integer constants are needed in this case. We used those in other classes (e.g. Slime) since they were convenient for delineating which sprites to draw. For this class, I think that strings are sufficient but am open to further discussion.

calebkrauter replied:
Sounds good! That is my mistake. I am a bit new to this type of mapping and mistook the "type" for an int. Good work!

tclaar replied:
No worries! Using named integers would still be a valid approach.

b1gd3vd0g commented on js/classes/entities/background/decoration.js:
```diff
@@ -0,0 +1,88 @@
+     * corner.
+     * 
+     * @param {string} type the type of Decoration to construct (must be a key to Decoration.PROPERTY_TABLE)
+     * @param {Vector} pos the position of the bottom left corner of the Decoration (in world space) 
```
I think, with a special position such as this, "pos" is a bit vague (although yes, the JSDoc does explain it). Consider bottomPos, posBL, something like that.

calebkrauter replied:
Agreed. Good idea.

tclaar replied:
Agreed, good suggestion. Thanks!

b1gd3vd0g reviewed (approved):
Left a couple nitpicky comments, but no reason not to approve. The decoration class is very interesting - I was wondering how you were gonna get that one done. The properties table is very clever.
Also, love the sprites! Excited to build our lil village

calebkrauter commented on js/classes/entities/background/decoration.js:
```diff
@@ -0,0 +1,88 @@
+
+        this.scale = properties.SCALE;
+        this.pos = new Vector(pos.x, pos.y - size.y * this.scale);
+        this.animator = new Animator(properties.SPRITESHEET, properties.SPRITESHEET_START_POS, 
```
Nice job making this use of Animator more readable. @tclaar

calebkrauter commented on js/classes/entities/background/decoration.js:
```diff
@@ -0,0 +1,88 @@
+            size, properties.FRAME_COUNT, properties.FRAME_DURATION);
+    };
+
+    /** A map containing decoration types (keys) and tables of their properties (values). */
```
I really like how you are doing this. @tclaar

calebkrauter commented on js/classes/entities/background/decoration.js:
```diff
@@ -0,0 +1,88 @@
+     * @returns {string} the file path to the spritesheet
+     */
+    static getSpritesheet(type) {
+        return Decoration.PROPERTY_TABLE[type].SPRITESHEET;
```
Good work!

calebkrauter commented on js/classes/entities/background/decoration.js:
```diff
@@ -0,0 +1,88 @@
+    /**
+     * Decoration update method. Does nothing.
+     */
+    update() {
```
Just wondering if we need an update function if it does nothing? @tclaar

tclaar replied:
The game engine will throw an error otherwise, as it will attempt to call `update()` on all entities that you add to it.

calebkrauter replied:
Interesting. Thanks for the explanation!

calebkrauter reviewed (approved):
Looks fantastic! Just a minor concern with some naming. Thanks for all your work!

tclaar committed:
Update Decoration class.
Add ability to throw errors for invalid type parameters.
Change constructor's position parameter name from pos to bottomLeftPos for clarity.

tclaar committed:
Merge pull request #81 from tclaar/main

Update Decoration class.
